@vates/nbd-client/index.js

```diff
@@ -83,19 +95,46 @@ module.exports = class NbdClient {
     // to tls during the handshake
     await this.#unsecureConnect()
     await this.#handshake()
-
+    this.#disconnected = false
     // reset internal state if we reconnected a nbd client
     this.#commandQueryBacklog = new Map()
     this.#waitingForResponse = false
   }
 
   async disconnect() {
+    if (this.#disconnected) {
+      return
+    }
+
     const buffer = Buffer.alloc(28)
     buffer.writeInt32BE(NBD_REQUEST_MAGIC, 0) // it is a nbd request
     buffer.writeInt16BE(0, 4) // no command flags for a disconnect
     buffer.writeInt16BE(NBD_CMD_DISC, 6) // we want to disconnect from nbd server
     await this.#write(buffer)
     await this.#serverSocket.destroy()
+    this.#serverSocket = undefined
+    this.#disconnected = true
+  }
+
+  async reconnect() {
+    if (this.#reconnectingPromise) {
+      return this.#reconnectingPromise
+    }
+    let resolveReconnect, rejectReconnect
+    this.#reconnectingPromise = new Promise((resolve, reject) => {
+      resolveReconnect = resolve
+      rejectReconnect = reject
+    })
```

Comment:
In my experience, it's usually easier to split this behavior in a separate function/method:

```js
async #reconnect() {}

#pReconnect
#clearPReconnect = () => {
  this.#pReconnect = undefined
}
reconnect() {
  if (this.#pReconnect === undefined) {
    this.#pReconnect = this.#reconnect()
    this.#pReconnect.then(this.#clearPReconnect, this.#clearPReconnect)
  }
  return this.#pReconnect
}
```

Or you can use `@vates/coalesce-calls`:

```js
async reconnect() {}
reconnect = coalesceCalls(this.reconnect)
```

Reply:
I am also using this promise to ensure we don't read block during reconnecting, thus I need to be able to access to the reconnect promise . If not I was running into race conditions with block query that were successfully sent between multiple connect / disconnect.  

Do you have a more elegant solution with an existing method ?

Reply:
I'm not sure I understand the issue, we can speak of this directly if you want to, but my first proposal supports this use case :slightly_smiling_face: